### PR TITLE
Enable ansible capsule install for dev environments

### PR DESCRIPTION
**Pull request by johnpmitsch**
To use this functionality, add the following configuration to your boxes.yaml,
changing the hostnames as needed

To setup a capsule with an existing development environment
------------------------------------------------------------

* Add to existing development server's configuration in boxes.yaml
```
  ansible:
    group: 'server'
```
* Add capsule with development server's name in server to boxes.yaml:

```
foocapsule:
  box: centos7
  ansible:
    playbook: 'playbooks/capsule-dev.yml'
    group: 'capsule'
    server: 'foo'
```
* ssh into existing development server and ```rails s```
* spin up new capsule ```vagrant up foocapsule```

To setup a new development environment with a capsule
-----------------------------------------------------

* setup boxes.yaml

```
foo:
  box: centos7
  shell: 'yum -y install ruby && cd /vagrant && ./setup.rb'
  options: --scenario=katello-devel
  installer: --katello-devel-github-username <your-github-name> --disable-system-checks
  ansible:
    group: 'server'

foocapsule:
  box: centos7
  ansible:
    playbook: 'playbooks/capsule-dev.yml'
    group: 'capsule'
    server: 'foo'
```
* ```vagrant up foo```
* ssh into foo and ```rails s```
* ```vagrant up foocapsule```

### CODE REVIEW

**Comment by johnpmitsch** [comment]
I think its possible to do some more refactoring, as the capsule and capsule-dev roles have repeating logic. Maybe we should have a capsule-base role and then split up the unique tasks to a capsule-dev and capsule-prod role?

**Comment by ehelms** [comment]
What about adding a basic centos7-capsule-dev box that points at the centos7-dev box and is configured?

File: playbooks/common-dev/tasks/repositories.yml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+---
+- name: 'Setup Epel Repository'
+  yum: name=http://dl.fedoraproject.org/pub/epel/epel-release-latest-{{ release }}.noarch.rpm
+- name: 'Setup Puppetlabs Repository'
+  yum: name=http://yum.puppetlabs.com/puppetlabs-release-el-{{ release }}.noarch.rpm
+- name: 'Setup Foreman Repository'
+  yum: name=http://yum.theforeman.org/releases/{{ foreman_version }}/el{{ release }}/x86_64/foreman-release.rpm
+- name: 'Setup Katello Repository'
+  yum: name=https://fedorapeople.org/groups/katello/releases/yum/{{ katello_version }}/katello/el{{ release }}/x86_64/katello-repos-latest.rpm
```
Comment (ehelms) [inline playbooks/common-dev/tasks/repositories.yml]: Do these differ from the common/tasks/repositories.yml file?

File: playbooks/common-dev/vars/main.yml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+---
+release: "{{ ansible_distribution_version.split('.')[0] }}"
+server_group: "server-{{ inventory_hostname }}"
+server: "{{ groups[server_group][0] }}"
+version_map:
+  'nightly': 'nightly'
+  '3.0': '1.11'
+katello_version: "nightly"
+foreman_version: "nightly"
```
Comment (ehelms) [inline playbooks/common-dev/vars/main.yml]: You could instead define this in `common/defaults/main.yml` and then override them:

```
---
 - hosts: server capsule
   roles:
     - {role: common, katello_version: 'nightly', foreman_version: 'nightly'}
     - etc_hosts
     - capsule-dev
```

File: plugin_library/sat-deploy [hunk 3]
```diff
@@ -0,0 +1 @@
+Subproject commit 53861905d7d13ff8d5e450992f1e150fc9618025
```
Comment (ehelms) [inline plugin_library/sat-deploy]: Unintended add?

File: playbooks/capsule-dev/tasks/main.yaml [hunk 4]
```diff
@@ -0,0 +1,52 @@
+---
```
Comment (ehelms) [inline playbooks/capsule-dev/tasks/main.yaml]: Can you identify what in this file differs from the capsule/tasks/main.yml?

File: playbooks/capsule-dev/tasks/install.yml [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    foreman-installer -v --scenario capsule
+    --foreman-proxy-trusted-hosts "{{ server_fqdn.stdout }}"
+    --foreman-proxy-trusted-hosts "{{ ansible_nodename }}"
+    --foreman-proxy-foreman-base-url "https://{{ server_fqdn.stdout }}"
+    --foreman-proxy-register-in-foreman true
+    --foreman-proxy-oauth-consumer-key "{{ oauth_consumer_key.stdout }}"
+    --foreman-proxy-oauth-consumer-secret "{{ oauth_consumer_secret.stdout }}"
+    --capsule-certs-tar "{{ capsule_certs_tar }}"
+    --capsule-parent-fqdn "{{ server_fqdn.stdout }}"
+    --capsule-pulp-oauth-secret "{{ pulp_oauth_secret.stdout }}"
```
Comment (ehelms) [inline playbooks/capsule-dev/tasks/install.yml]: Can you identify what in this file differs from the capsule/tasks/main.yml?

File: playbooks/capsule-dev/tasks/main.yaml (hunk 4 above)
Comment (johnpmitsch) [inline playbooks/capsule-dev/tasks/main.yaml]: just the file paths and the permissions change, here is the diff

``` 
19c20
<   shell: grep oauth_consumer_key /home/vagrant/foreman/config/settings.yaml | cut -d ' ' -f 2
---
>   shell: grep oauth_consumer_key /etc/foreman/settings.yaml | cut -d ' ' -f 2
24c25
<   shell: grep oauth_consumer_secret /home/vagrant/foreman/config/settings.yaml | cut -d ' ' -f 2
---
>   shell: grep oauth_consumer_secret /etc/foreman/settings.yaml | cut -d ' ' -f 2
41,44d41
< 
< - name: 'Change cert permissions'
<   shell: chmod 755 -R /etc/pki/katello/private
<   delegate_to: "{{ server }}"
```
maybe we could make a file base path variable and use that?

File: playbooks/capsule-dev/tasks/install.yml (hunk 5 above)
Comment (johnpmitsch) [inline playbooks/capsule-dev/tasks/install.yml]: exactly the same

**Comment by johnpmitsch** [comment]
added the base boxes

File: playbooks/common-dev/tasks/repositories.yml (hunk 1 above)
Comment (johnpmitsch) [inline playbooks/common-dev/tasks/repositories.yml]: also exactly the same

File: config/base_boxes.yaml [hunk 6]
```diff
@@ -66,6 +66,8 @@ boxes:
   centos7-bats:
     box: centos7
     <<: *bats_shell
+    ansible:
+      group: 'server'
```
Comment (ehelms) [inline config/base_boxes.yaml]: Wrong box I think.

File: playbooks/capsule-dev/tasks/main.yaml (hunk 4 above)
Comment (ehelms) [inline playbooks/capsule-dev/tasks/main.yaml]: We could turn the parts that are the same into their own task files and the parts that are different, vary those depending on a variable passed in.

**Comment by johnpmitsch** [comment]
@ehelms refactored, still getting errors on the install, marking this WIP

**Comment by stbenjam** [comment]
Needs a rebase as well

**Comment by johnpmitsch** [comment]
@ehelms @stbenjam updated this, it is currently working for one capsule, I need to still test with multiple capsules, and make sure centos6 works. I'll comment back when I do.

**Comment by ehelms** [comment]
To test your scenario, I needed to change this:

```
diff --git a/playbooks/roles/common/vars/main.yml b/playbooks/roles/common/vars/main.yml
index 6d51aab..39a31fd 100644
--- a/playbooks/roles/common/vars/main.yml
+++ b/playbooks/roles/common/vars/main.yml
@@ -3,5 +3,5 @@ release: "{{ ansible_distribution_version.split('.')[0] }}"
 version_map:
   'nightly': 'nightly'
   '3.0': '1.11'
-katello_version: "{{ inventory_hostname.split('-')[-1] }}"
+katello_version: "{{ inventory_host.split('-')[-1] if inventory_hostname.split('-')[-1] in version_map.keys() else 'nightly' }}"
 foreman_version: "{{ version_map[katello_version] }}"
```

**Comment by johnpmitsch** [comment]
@ehelms updated. tested with a rails server running on the main server, the install finished successfully

**Comment by johnpmitsch** [comment]
@ehelms now updated to start the rails server on the main katello server. I don't see a need to check for a rails server running as `rails s` will just exit if one is already running.

**Comment by johnpmitsch** [comment]
Tested this using ```vagrant up foo && vagrant up foo-capsule``` which completed successfully and then was able to spin up another capsule successfully on the same main server.

File: playbooks/roles/capsule/tasks/install.yml [hunk 7]
```diff
@@ -47,6 +47,11 @@
     src: "/tmp/{{ inventory_hostname }}{{ capsule_certs_tar }}"
     dest: ~/
 
+- name: 'Change cert permissions'
+  shell: chmod 755 -R /etc/pki/katello/private
+  delegate_to: "{{ capsule_server }}"
+  when: devel == true
```
Comment (ehelms) [inline playbooks/roles/capsule/tasks/install.yml]: Why is this needed?

File: playbooks/roles/capsule/tasks/devel_install.yml [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+  args:
+    chdir: /home/vagrant/foreman
+  delegate_to: "{{ capsule_server }}"
+  async: 1000
+  poll: 0
+  become: yes
+  become_user: vagrant
+
+- name: 'Install foreman-installer-katello'
+  yum: name="foreman-installer-katello"
```
Comment (ehelms) [inline playbooks/roles/capsule/tasks/devel_install.yml]: Why do we need to install this on an already existing dev server?

File: playbooks/roles/capsule/tasks/devel_install.yml [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  delegate_to: "{{ capsule_server }}"
+
+- name: 'Add group foreman'
```
Comment (ehelms) [inline playbooks/roles/capsule/tasks/devel_install.yml]: What is this intended for?

File: playbooks/roles/capsule/tasks/devel_install.yml [hunk 10]
```diff
@@ -0,0 +1,19 @@
+- name: 'start rails server'
+  command: /home/vagrant/.rvm/bin/rvm ruby-2.2.4 do rails s
```
Comment (ehelms) [inline playbooks/roles/capsule/tasks/devel_install.yml]: We should probably set this to just ruby if possible to avoid specific version constraints.

**Comment by ehelms** [comment]
I still ended up without oauth settings being configured and thus a 401 on my capsule install.

File: playbooks/roles/capsule/tasks/install.yml (hunk 7 above)
Comment (johnpmitsch) [inline playbooks/roles/capsule/tasks/install.yml]: Just something I have had to do in the dev environment for a successful capsule install

File: playbooks/roles/capsule/tasks/devel_install.yml (hunk 9 above)
Comment (johnpmitsch) [inline playbooks/roles/capsule/tasks/devel_install.yml]: Another thing that is required for some reason when installing a capsule in a dev environment

File: playbooks/roles/capsule/tasks/devel_install.yml (hunk 10 above)
Comment (johnpmitsch) [inline playbooks/roles/capsule/tasks/devel_install.yml]: @ehelms I have to recall my findings when I wrote this, but I think the rvm and ansible relationship can be tricky and this was a workaround I found to use an rvm environment. I'll investigate further today. There is a [plugin](https://github.com/rvm/rvm1-ansible) that we may find useful

**Comment by johnpmitsch** [comment]
@ehelms removed the unnecessary installation of foreman-installer-katello on the main katello server and added "become = yes" to both capsule and capsule-dev playbooks. I tested spinning up a  dev capsule and a server with no issue.

**Comment by ehelms** [comment]
@johnpmitsch I'll work on testing this, in the meantime this will need to be rebased with master.

**Comment by johnpmitsch** [comment]
@ehelms appreciate it, rebashed and pushed up again

File: playbooks/capsule-dev.yml [hunk 11]
```diff
@@ -0,0 +1,10 @@
+---
+- hosts: server capsule
+  become: yes
+  vars:
+    capsule_server_group: "server-{{ inventory_hostname }}"
+    capsule_server: "{{ groups[capsule_server_group][0] }}"
+  roles:
+    - {role: common, katello_version: 'nightly', foreman_version: 'nightly'}
```
Comment (ehelms) [inline playbooks/capsule-dev.yml]: I'd recommend switching this to:

```
  - foreman_repositories
  - katello_repositories
```

**Comment by ehelms** [comment]
I am finding that capsule-certs-generate is not vailable on a devel install because it is required by foreman-installer-katello and not foreman-installer-katello-devel. I've opened a PR to fix this -- https://github.com/Katello/katello-packaging/pull/263

File: playbooks/roles/capsule/tasks/install.yml (hunk 7 above)
Comment (zjhuntin) [inline playbooks/roles/capsule/tasks/install.yml]: A small nitpick for this, you can use the file module instead of chmod... only brought it up because Ansible did first. I think it'd be something along the lines of:
`file: path='/etc/pki/katello/private' mode=0775`

http://docs.ansible.com/ansible/file_module.html

**Comment by zjhuntin** [comment]
Other than my small comment this is working quite well for me. Although I guess there are still some concerns regarding the katello-packaging PR.

**Comment by johnpmitsch** [comment]
@ehelms Do we need to wait on the katello-packaging PR to ACK this?

**Comment by ehelms** [comment]
@johnpmitsch if this doesnt work without the packaging change I would say yes wait

**Comment by johnpmitsch** [comment]
@ehelms It has been working fine for myself and @zjhuntin

**Comment by johnpmitsch** [comment]
@ehelms 
With the latest rebase I'm hitting an error here:

```
TASK [capsule : Generate Certs] ************************************************
fatal: [chickencapsule -> None]: FAILED! => {"changed": false, "cmd": "capsule-certs-generate --capsule-fqdn chickencapsule.example.com --certs-tar /root/chickencapsule.example.com.tar.gz", "failed": true, "msg": "[Errno 2] No such file or directory", "rc": 2}

cmd: capsule-certs-generate --capsule-fqdn chickencapsule.example.com --certs-tar /root/chickencapsule.example.com.tar.gz

msg: [Errno 2] No such file or directory
 [ERROR]: Could not create retry file '/home/jomitsch/.ansible/retry-files
/capsule-dev.retry'. The error was: [Errno 2] No such file or directory:
'/home/jomitsch/.ansible/retry-files/capsule-dev.retry'
```

any ideas what could be causing this?

**Comment by ehelms** [comment]
@johnpmitsch comes from that retry-files directory missing

**Comment by johnpmitsch** [comment]
@ehelms 

I realized the previous error I posted is because of not having install 'foreman-installer-katello' step (which installs capsule-certs-generate) 

Since this is the workflow to install a capsule in dev now, I would say having this step should not hold up this PR. Once https://github.com/Katello/katello-packaging/pull/263 is merged we can remove the step

File: playbooks/roles/capsule/tasks/install.yml [hunk 12]
```diff
@@ -44,6 +45,11 @@
     src: "/tmp/{{ inventory_hostname }}{{ capsule_certs_tar }}"
     dest: ~/
 
+- name: 'Change cert permissions'
+  file: path='/etc/pki/katello/private' mode=0775
+  delegate_to: "{{ capsule_server }}"
+  when: devel == true
```
Comment (ehelms) [inline playbooks/roles/capsule/tasks/install.yml]: Why is this needed?

File: playbooks/roles/capsule/tasks/devel_install.yml [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+  become: yes
+  become_user: vagrant
+
+- name: 'Add group foreman'
+  group: name=foreman state=present
+  delegate_to: "{{ capsule_server }}"
+
+- name: 'Install foreman-installer-katello'
+  yum: name="foreman-installer-katello"
+  delegate_to: "{{ capsule_server }}"
```
Comment (ehelms) [inline playbooks/roles/capsule/tasks/devel_install.yml]: I had missed that you added this here.

File: playbooks/roles/capsule/tasks/install.yml (hunk 12 above)
Comment (johnpmitsch) [inline playbooks/roles/capsule/tasks/install.yml]: @ehelms you already asked :smile: [here](https://git.io/vitJx) Its just something that we have to do in dev currently

File: playbooks/roles/capsule/tasks/devel_install.yml (hunk 10 above)
Comment (ehelms) [inline playbooks/roles/capsule/tasks/devel_install.yml]: We should probably stop the server at the end?

File: playbooks/roles/capsule/tasks/devel_install.yml (hunk 13 above)
Comment (johnpmitsch) [inline playbooks/roles/capsule/tasks/devel_install.yml]: I took it out since you had the packaging PR up, but added it back in since I don't think we should hold this PR for that one

File: playbooks/roles/capsule/tasks/devel_install.yml (hunk 10 above)
Comment (johnpmitsch) [inline playbooks/roles/capsule/tasks/devel_install.yml]: @ehelms The only issue I see with that is if the developer has a server already running and uses this automation, it would kill their running server which may not be what they want. I agree if it isn't running and we start it in this playbook, we should stop it as well, if there is a way to just do that.

Comment (johnpmitsch) [inline playbooks/roles/capsule/tasks/devel_install.yml]: @ehelms also we can do this to avoid ruby constraints

```/home/vagrant/.rvm/bin/rvm $(rvm current) do rails s```

Comment (ehelms) [inline playbooks/roles/capsule/tasks/devel_install.yml]: Wouldn't this in theory error if they already have one running?

Comment (johnpmitsch) [inline playbooks/roles/capsule/tasks/devel_install.yml]: It just exits saying "a server is already running" I am pretty sure I have had one running when I ran this playbook, but will test it again to be sure.

**Comment by beav** [comment]
This worked for me with Puppet 3. I hit an unrelated error with Puppet 4, but it would not impact this PR at all.

ACK

**Comment by johnpmitsch** [comment]
Leaving it up to @ehelms for the final ACK

**Comment by ehelms** [comment]
ACK - thanks @johnpmitsch !
